examples_tests/02.ComputeShader/main.cpp

```diff
@@ -27,11 +38,584 @@ void main()
+	core::smart_refctd_ptr<video::IGPUFence> frameFences[FRAMES_IN_FLIGHT];
+	for (uint32_t i = 0u; i < FRAMES_IN_FLIGHT; ++i)
+	{
+		acquireSemaphores[i] = device->createSemaphore();
+		releaseSemaphores[i] = device->createSemaphore();
+		frameFences[i] = device->createFence(video::IGPUFence::E_CREATE_FLAGS::ECF_SIGNALED_BIT);
+	}
+
+	// Record commands in commandBuffers here
+	for (uint32_t i = 0u; i < swapchainImageCount; ++i)
+	{
+		video::IGPUCommandBuffer::SImageMemoryBarrier undefToComputeTransitionBarrier;
+		undefToComputeTransitionBarrier.barrier.srcAccessMask = asset::EAF_TRANSFER_READ_BIT;
+		undefToComputeTransitionBarrier.barrier.dstAccessMask = asset::EAF_SHADER_READ_BIT;
+		undefToComputeTransitionBarrier.oldLayout = asset::EIL_UNDEFINED;
+		undefToComputeTransitionBarrier.newLayout = asset::EIL_GENERAL;
+		undefToComputeTransitionBarrier.srcQueueFamilyIndex = ~0u;
+		undefToComputeTransitionBarrier.dstQueueFamilyIndex = ~0u;
+		undefToComputeTransitionBarrier.image = *(swapchainImages.begin() + i);
+		undefToComputeTransitionBarrier.subresourceRange.aspectMask = asset::IImage::EAF_COLOR_BIT;
+		undefToComputeTransitionBarrier.subresourceRange.baseMipLevel = 0u;
+		undefToComputeTransitionBarrier.subresourceRange.levelCount = 1u;
+		undefToComputeTransitionBarrier.subresourceRange.baseArrayLayer = 0u;
+		undefToComputeTransitionBarrier.subresourceRange.layerCount = 1u;
+
+		video::IGPUCommandBuffer::SImageMemoryBarrier computeToPresentTransitionBarrier;
+		computeToPresentTransitionBarrier.barrier.srcAccessMask = asset::EAF_SHADER_WRITE_BIT;
+		computeToPresentTransitionBarrier.barrier.dstAccessMask = static_cast<asset::E_ACCESS_FLAGS>(0);
+		computeToPresentTransitionBarrier.oldLayout = asset::EIL_GENERAL;
+		computeToPresentTransitionBarrier.newLayout = asset::EIL_PRESENT_SRC_KHR;
+		computeToPresentTransitionBarrier.srcQueueFamilyIndex = ~0u;
+		computeToPresentTransitionBarrier.dstQueueFamilyIndex = ~0u;
+		computeToPresentTransitionBarrier.image = *(swapchainImages.begin() + i);
+		computeToPresentTransitionBarrier.subresourceRange.aspectMask = asset::IImage::EAF_COLOR_BIT;
+		computeToPresentTransitionBarrier.subresourceRange.baseMipLevel = 0u;
+		computeToPresentTransitionBarrier.subresourceRange.levelCount = 1u;
+		computeToPresentTransitionBarrier.subresourceRange.baseArrayLayer = 0u;
+		computeToPresentTransitionBarrier.subresourceRange.layerCount = 1u;
+
+		commandBuffers[i]->begin(0);
+
+		// Todo(achal): The fact that this pipeline barrier is solely on a compute queue might
+		// affect the srcStageMask. More precisely, I think, for some reason, that
+		// VK_PIPELINE_STAGE_TRANSFER_BIT shouldn't be specified in compute queue
+		// but present queue (or transfer queue if theres one??)
+		commandBuffers[i]->pipelineBarrier(asset::EPSF_TRANSFER_BIT,
+			asset::EPSF_COMPUTE_SHADER_BIT, 0, 0u, nullptr, 0u, nullptr, 1u,
+			&undefToComputeTransitionBarrier);
```

Comment:
@Crisspl correct me if I'm wrong, but cpu2gpu object converter should do big fat barriers in its last submit and handle the queue ownership transfer AFAIK

@achalpandeyy you shouldn't be worrying about barriering, transfering ownership from transfer queue to compute, and transitioning out of TRANSFER_DST to SHADER_READ_ONLY_OPTIMAL

only if you need to transition from the SHADER_READ_ONLY_OPTIMAL assumed by cpu2gpu converter for all textures, you need a transition, but only once! So dont bake it into a reusable per-frame commandbuffer, just fire off a single simple submit before you start rendering.

Reply:
the barrier that needs to be here is TOP_OF_PIPE->COMPUTE

Reply:
I just learnt swapchain images start out in undefined, so you need a minidispatch if you transitioned from not UNDEFINED to GENERAL

Reply:
I guess this is no longer relevant? Since these barriers are for layout transition of swapchain images (UNDEFINED --> GENERAL --> PRESENT_SRC), I don't see how the asset converter would help me here?

Reply:
I think its impossible to untangle the original context after so many months, we'll just deal with validation errors when they arise.